x/il/keeper/abci.go

```diff
@@ -0,0 +1,168 @@
+package keeper
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/armon/go-metrics"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	bridgetypes "github.com/althea-net/peggy/module/x/peggy/types"
+
+	"github.com/peggyjv/sommelier/x/il/types"
+	"github.com/peggyjv/sommelier/x/il/types/contract"
+	oracletypes "github.com/peggyjv/sommelier/x/oracle/types"
+)
+
+// TODO: the stoploss position should be recreated if the tx on ethereum timeouts or fails.
+
+// EndBlocker is called at the end of every block
+func (k Keeper) EndBlocker(ctx sdk.Context) {
+	defer telemetry.ModuleMeasureSince(types.ModuleName, time.Now(), telemetry.MetricKeyEndBlocker)
+
+	// check if the latest eth block height
+	ethHeight := k.ethBridgeKeeper.GetLastObservedEthereumBlockHeight(ctx)
+	if ethHeight.EthereumBlockHeight == 0 {
+		k.Logger(ctx).Debug("tracked ethereum height is 0")
+		return
+	}
+
+	params := k.GetParams(ctx)
+
+	invalidationID := k.GetInvalidationID(ctx)
+
+	// variable for the original value to check if we need to store the new invalidationID value
+	originalInvalidationID := invalidationID
+
+	pairMap := make(map[string]*oracletypes.UniswapPair)
+
+	k.IterateStoplossPositions(ctx, func(address sdk.AccAddress, stoploss types.Stoploss) (stop bool) {
```

Comment:
It would be interesting to do some benchmarking on this later on. Depending on the usage we could run into similar issue as kava with validators missing blocks because they have too much in the endblock/beginblock.

Reply:
Oh wait until you see the Peggy module endblocker, that one is large

Reply:
![](https://media0.giphy.com/media/lvzdeWk12qjmM/giphy.gif)

Reply:
Sommelier is expected to be a very resource intensive process for validators to run.